### PR TITLE
add metrics crate

### PR DESCRIPTION
**Pull request by rc-andres**
follow-up from: https://github.com/redcanaryco/oxidebpf/pull/47

This lets users of `oxidebpf` to optionally elect into receiving metrics via the `metrics` crate for size of the mmap and how filled it is right before being read. This should let users of `oxidebpf` determine how their perf buffers are being used. In the future we may add more metrics under the same feature flag.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -16,7 +16,7 @@ authors = [
     "Andrés Medina <andres.medina@redcanary.com>",
     "Vince Bundage <vince.bundage@redcanary.com>"
 ]
-edition = "2021"
+edition = "2018"
```
Comment (RafaelOrtizRC) [inline Cargo.toml:19]: I think we have some 2021-isms that are causing this to fail CI. Is 2018 necessary here?

Specifically, the `thread-priority` tests are failing with:
```
no method named `try_into` found for type `u8` in the current scope
```

If I raise the edition back to 2021, everything passes.

**Review by RafaelOrtizRC — CHANGES_REQUESTED** [review]
Do we need to list this as a feature? 

```
[features]
log_buf = []
rootless_blueprints = []
metrics = []
```

There's also a CI issue re: 2018 vs 2021 edition.

File: Cargo.toml (hunk 1 above)
Comment (RafaelOrtizRC) [reply Cargo.toml:19]: Oh wow, beat me to it.

Comment (rc-andres) [reply Cargo.toml:19]: I fixed the issue by adding the import in a sneaky force push 😁

**Review by RafaelOrtizRC — APPROVED** [review]
This is going to be super helpful!

File: README.md [hunk 2]
```diff
@@ -118,3 +118,29 @@ Note: some tests will require root privileges to pass. Other tests require a sin
...
+* `mmap.buffer_slack_kb`: A histogram on how much memory (in KB) are
+  unused. This is emitted right after polling the perf map file
+  descriptors but before reading from them. This is only emitted by
+  perfmaps that woke up during the epoll, if a perfmap never gets data
+  then it will not be reported by this metric. The map name and cpu
+  are reported as labels.
+
+* `poller.channel.full`: A counter of how many times we tried to send
+  a dropped message across the user provided channel for perfmaps but
+  failed to it being full. The map name is reported as a label.
```
Comment (pinealservo) [inline README.md:140]: I think you forgot to type a word, i.e. "failed to it being full" -> "failed due to it being full"

File: src/maps/mod.rs [hunk 3]
```diff
@@ -299,33 +299,49 @@ struct PerfEventIterator<'a> {
 
     // technically not needed but it gives us the lifetime we need to
```
Comment (pinealservo) [inline src/maps/mod.rs:300]: Although it wasn't technically necessary to store a reference, I think it *was* technically necessary to have some member associated with the lifetime parameter.  So I'd remove the "technically not needed but" part from the comment, and the rest is a fine reminder of what the `PhantomData` is for.

**Review by pinealservo — APPROVED** [review]
Looking good, aside from a minor nit about a comment.

**Pull request merged**